### PR TITLE
fix memory leak issue with the FinalizerRunnable

### PR DESCRIPTION
**Pull request by nhachicha**
@realm/java

### CODE REVIEW

File: realm/src/androidTest/java/io/realm/RealmTest.java [hunk 1]
```diff
@@ -1674,4 +1678,32 @@ public void testMutableMethodsOutsideWriteTransactions() throws JSONException, I
...
+        Field fieldIsFinalizerStarted = Realm.class.getDeclaredField("isFinalizerStarted");
+        fieldIsFinalizerStarted.setAccessible(true);
+        boolean isFinalizerStarted = fieldIsFinalizerStarted.getBoolean(null);
+        assertTrue(isFinalizerStarted);
+
+        //insert some rows, then give the FinalizerRunnable some time to cleanup
+        populateTestRealm(testRealm, 20);
+
+        final int MAX_WAITING_RETRY = 5;
+        int nbRetry = 0;
```
Comment (emanuelez) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: What's the nb prefix about?

File: realm/src/androidTest/java/io/realm/RealmTest.java [hunk 2]
```diff
@@ -1674,4 +1678,32 @@ public void testMutableMethodsOutsideWriteTransactions() throws JSONException, I
...
+        while (references.size() > 0 && nbRetry < MAX_WAITING_RETRY) {
+            SystemClock.sleep(5);//5ms
```
Comment (emanuelez) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: Space after `;`

**Comment by emanuelez** [comment]
:+1: apart from minor comments

File: realm/src/androidTest/java/io/realm/RealmTest.java (hunk 1 above)
Comment (nhachicha) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: it will try 5 times then fail, instead of waiting forever :)

Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: Isn't `no` a more common prefix for number?

File: realm/src/androidTest/java/io/realm/RealmTest.java [hunk 3]
```diff
@@ -1674,4 +1678,32 @@ public void testMutableMethodsOutsideWriteTransactions() throws JSONException, I
...
+        fieldReferences.setAccessible(true);
+        Map<Reference<?>, Boolean> references = (Map<Reference<?>, Boolean>) fieldReferences.get(null);
+        assertNotNull(references);
+
+        Field fieldIsFinalizerStarted = Realm.class.getDeclaredField("isFinalizerStarted");
+        fieldIsFinalizerStarted.setAccessible(true);
+        boolean isFinalizerStarted = fieldIsFinalizerStarted.getBoolean(null);
+        assertTrue(isFinalizerStarted);
+
+        //insert some rows, then give the FinalizerRunnable some time to cleanup
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: Non-deterministic tests, yay. Let's hope this doesn't bite us hard, and no I don't have a better idea :)

Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: Point of fact, CI just failed :) Devices are slower than emulators, so maybe just raising the timer or hitting System.gc() a few more times might help

**Comment by cmelchior** [comment]
Changelog missing, and apart from smaller comments :+1:

File: realm/src/androidTest/java/io/realm/RealmTest.java (hunk 3 above)
Comment (nhachicha) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: I can call `RealmCore.gcGuaranteed` but it was blocking the test so far

File: realm/src/androidTest/java/io/realm/RealmTest.java (hunk 1 above)
Comment (bmunkholm) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: `retries` could do

File: realm/src/androidTest/java/io/realm/RealmTest.java (hunk 3 above)
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: Yes, my point is just that this unit test is pending on the behavior of the runtime and it is entirely possible that ART will not GC if not memory pressed no matter how long we call System.gc(). Maybe you actually need to test a reasonable number of entries (say 5K) and just check that the GC thread has a lower number of entries than that?

File: realm/src/androidTest/java/io/realm/RealmTest.java [hunk 4]
```diff
@@ -1674,4 +1679,45 @@ public void testMutableMethodsOutsideWriteTransactions() throws JSONException, I
...
+            SystemClock.sleep(TimeUnit.SECONDS.toMillis(1)); //1s
+            numberbRetries++;
+            System.gc();
+        }
+
+        // we can't guarantee that all references have been GC'd but we should detect a decrease
+        boolean isDecreasing = references.size() < totalNumberOfReferences;
+        if (!isDecreasing) {
+            fail("FinalizerRunnable is not closing all native resources");
+
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: whitespace

File: realm/src/androidTest/java/io/realm/RealmTest.java [hunk 5]
```diff
@@ -1674,4 +1679,45 @@ public void testMutableMethodsOutsideWriteTransactions() throws JSONException, I
...
+        } else {
+            android.util.Log.d(RealmTest.class.getName(), "FinalizerRunnable freed : "
+                    + (totalNumberOfReferences - references.size()) + " out of " + totalNumberOfReferences);
+        }
+
```
Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: whitespace

File: changelog.txt [hunk 6]
```diff
@@ -1,3 +1,6 @@
+0.81.1
+ * Fixed a memory leak issue with the FinalizerRunnable (PhantomReference)
```
Comment (cmelchior) [inline changelog.txt]: This is very internal. Perhaps rephrase to "Fixed memory leak causing Realm to never release Realm objects." or something that would make sense from a user stand point.

File: realm/src/androidTest/java/io/realm/RealmTest.java [hunk 7]
```diff
@@ -1674,4 +1679,45 @@ public void testMutableMethodsOutsideWriteTransactions() throws JSONException, I
...
+        // we have 8 reference so far let's add more
+        final int numberOfPopulateTest = 10000;
+        final int totalNumberOfReferences = 8 + 20 * 2 * numberOfPopulateTest;
+
+        for (int i=0; i<numberOfPopulateTest; i++) {
+            populateTestRealm(testRealm, 20);
+        }
+
+        final int MAX_GC_RETRIES = 5;
+        int numberbRetries = 0;
```
Comment (bmunkholm) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: remove a `b` after `number`

File: realm/src/androidTest/java/io/realm/RealmTest.java [hunk 8]
```diff
@@ -1674,4 +1679,45 @@ public void testMutableMethodsOutsideWriteTransactions() throws JSONException, I
...
+        fieldIsFinalizerStarted.setAccessible(true);
+        boolean isFinalizerStarted = fieldIsFinalizerStarted.getBoolean(null);
+        assertTrue(isFinalizerStarted);
+
+        //insert some rows, then give the FinalizerRunnable some time to cleanup
+        // we have 8 reference so far let's add more
+        final int numberOfPopulateTest = 10000;
+        final int totalNumberOfReferences = 8 + 20 * 2 * numberOfPopulateTest;
+
+        for (int i=0; i<numberOfPopulateTest; i++) {
```
Comment (bmunkholm) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: add spacing around `=` and `<`.

(Gold points to the first that get our coding style checking automated...)

Comment (nhachicha) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: do we have a custom CheckStyle file I can apply ?

Comment (cmelchior) [inline realm/src/androidTest/java/io/realm/RealmTest.java]: No, but we have been talking about making one for a while now :)

**Comment by cmelchior** [comment]
:+1: Great we tracked this down before hitting the big release button :shipit:

**Pull request merged**